scala_proto/private/scala_proto_aspect.bzl

```diff
@@ -122,27 +133,60 @@ def _compile_sources(ctx, toolchain, proto, src_jars, deps):
         runtime_deps = compile_deps,
     )
 
-####
-# This is applied to the DAG of proto_librarys reachable from a deps
-# or a scalapb_scala_library. Each proto_library will be one scalapb
-# invocation assuming it has some sources.
-def _scalapb_aspect_impl(target, ctx):
+def _phase_proto_provider(ctx, p):
+    return p.target[ProtoInfo]
+
+def _phase_deps(ctx, p):
+    return [d[ScalaProtoAspectInfo].java_info for d in ctx.rule.attr.deps]
+
+def _phase_generate_and_compile(ctx, p):
+    proto = p.proto_info
+    deps = p.deps
+    stamp_label = p.stamp_label
     toolchain = ctx.toolchains["@io_bazel_rules_scala//scala_proto:toolchain_type"]
-    proto = target[ProtoInfo]
-    deps = [d[ScalaPBAspectInfo].java_info for d in ctx.rule.attr.deps]
 
     if proto.direct_sources and _code_should_be_generated(ctx, toolchain):
         src_jars = _generate_sources(ctx, toolchain, proto)
-        java_info = _compile_sources(ctx, toolchain, proto, src_jars, deps)
-        return [ScalaPBAspectInfo(java_info = java_info)]
+        java_info = _compile_sources(ctx, toolchain, proto, src_jars, deps, stamp_label)
+        return java_info
     else:
         # this target is only an aggregation target
-        return [ScalaPBAspectInfo(java_info = java_common.merge(deps))]
+        return java_common.merge(deps)
+
+def _phase_aspect_provider(ctx, p):
+    return struct(
+        external_providers = {
+            "ScalaProtoAspectInfo": ScalaProtoAspectInfo(java_info = p.generate_and_compile),
+        },
+    )
+
+def _phase_stamp_label(ctx, p):
+    rule_label = str(p.target.label)
+    toolchain = ctx.toolchains["@io_bazel_rules_scala//scala_proto:toolchain_type"]
 
-scalapb_aspect = aspect(
-    implementation = _scalapb_aspect_impl,
-    attr_aspects = ["deps"],
-    incompatible_use_toolchain_transition = True,
+    if toolchain.stamp_by_convention and rule_label.endswith("_proto"):
+        return rule_label.rstrip("_proto") + "_scala_proto"
```

Comment:
@liucijus 
Does starlark's `rstrip` differ from python's one?
In python it strips all characters from argument in any order:
```
>>> "otorp_proto".rstrip("_proto")
''
```

Reply:
@darl thanks for pointing this out - yes this is incorrect implementation